WI #557: Preview - Refinement of FOCUS Columns Normative Requirements (#664)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: supporting_content/guidelines/normative_requirements_overview.md

* This column MUST be of type String and MUST NOT contain null values.
* BillingAccountId MUST be a globally unique identifier within a provider.

## Column: Billing Account Name

### **Billing Account Name v.1.2 (Simplified Refinement)**

The BillingAccountName column adheres to the following requirements:

* BillingAccountName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* BillingAccountName MUST be of type String.
* BillingAccountName MUST conform to [String Handling](#stringhandling) requirements.
* BillingAccountName MUST NOT be null when the provider supports assigning a display name for the *billing account*.
* BillingAccountName MUST be unique within a customer.

### **Billing Account Name v.1.2 (Technical Refinement)**

The BillingAccountName column adheres to the following requirements:

* BillingAccountName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* BillingAccountName MUST be of type String.
* BillingAccountName MUST conform to [String Handling](#stringhandling) requirements.
* BillingAccountName MUST NOT be null when the provider supports assigning a display name for the *billing account*.

> * BillingAccountName MUST be unique within a customer.

Author: Skipped 'when BillingAccountName is not null' as it sounds awkward and felt unnecessary

Reviewer: Agreed!

* CapacityReservationStatus MUST be one of the allowed values.
* CapacityReservationStatus MUST label all unused *capacity reservation* charges and MUST label used *capacity reservation* charges if the provider supports it.

## Column: Charge Category

### **Charge Category v.1.2 (Simplified Refinement)**

The ChargeCategory column adheres to the following requirements:

* ChargeCategory MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ChargeCategory MUST be of type String.
* ChargeCategory MUST NOT be null.
* ChargeCategory MUST be one of the allowed values.

### **Charge Category v.1.2 (Technical Refinement)**

The ChargeCategory column adheres to the following requirements:

* ChargeCategory MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ChargeCategory MUST be of type String.
* ChargeCategory MUST NOT be null.
* ChargeCategory MUST be one of the allowed values.

### **Charge Category v.1.1 (Original)**

The ChargeCategory column adheres to the following requirements:

* The ChargeCategory column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) and MUST NOT be null.

> * This column is of type String and MUST be one of the allowed values.

Reviewer: In `CapacityReservationStatus`, you augmented the definition to include when to use the allowed values.  Here, you did not.  Should we try to make that more consistent, or is it okay not to?  In any case, it would be good if the guidelines spoke to this.

Author: I agree that we should take a consistent approach for all columns (and explicitly mention this in the guidelines). However, in this iteration, I chose to stick to the original specifications rather than expanding the normative requirements - **assessed that we wouldn't be able to cover this within version 1.2.**

ChargeCategory was one of the first columns we introduced, and at that time, the Normative Requirements only stated that '<ColumnId> MUST be one of the allowed values' (the same applies to CommitmentDiscountStatus, CommitmentDiscountCategory, etc.). At some point, we began adding more details to the Normative Requirements.

Another example: for CommitmentDiscountQuantity, the normative requirements specify the expected results (business context) when summing values in specific use case scenarios (Purchase, Usage, etc.). In contrast, no such business requirements were included for PricingQuantity (1.0-preview).

Reviewer: Discussed and aligned.  We don't state when to use each value in this revision because we didn't state that in 1.1 either.  Potentially we could augment requirements to declare when to use each, but that would be a substantial change to the definition.  Resolving with no action.

## Column: Charge Class

## Column: Charge Description

## Column: Charge Frequency

## Column: Charge Period End

## Column: Charge Period Start

## Column: Commitment Discount Category

## Column: Commitment Discount ID

## Column: Commitment Discount Name

## Column: Commitment Discount Status

## Column: Commitment Discount Type

## Column: Commitment Discount Quantity

### **CommitmentDiscountQuantity v.1.2 (Simplified Refinement)**

The CommitmentDiscountQuantity column adheres to the following requirements:

* CommitmentDiscountQuantity MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *commitment discounts*.
* CommitmentDiscountQuantity MUST be of type Decimal.
* CommitmentDiscountQuantity MUST conform to [Numeric Format](#numericformat) requirements.
* CommitmentDiscountQuantity nullability is defined as follows:

>   * When ChargeCategory is "Usage" or "Purchase" and CommitmentDiscountId is not null, CommitmentDiscountQuantity adheres to the following additional requirements:

Reviewer: The use of "adheres to the following additional requirements" here feels out of place.  Would it be okay to simplify and go with:

```suggestion
  * When ChargeCategory is "Usage" or "Purchase" and CommitmentDiscountId is not null:
```

If that's not enough detail, is there a rephrasing we could put here?  Perhaps:

```suggestion
  * When ChargeCategory is "Usage" or "Purchase" and CommitmentDiscountId is not null, CommitmentDiscountQuantity nullability is defined as follows:
```

Author: I followed the agreed-upon template but agree with the suggestion to simplify it and apply the same approach consistently across all columns, not just this one (I assume that was your intention as well).

Adding to the list of Discussion Topics.

Reviewer: Discussed.  Irena to determine where else to apply this and will proceed globally.

Reviewer: Irena to have mergeable draft PR for tomorrow (Feb 25) with comment of the issue.

Author: Alternative provided in the PR #798. Guideline outlined in PR #787

>   * When ChargeCategory is "Usage" or "Purchase" and CommitmentDiscountId is not null, CommitmentDiscountQuantity adheres to the following additional requirements:

Reviewer: I think we also have to be a bit more specific with our language here, because the logic of AND and OR is implied here, and we need to be explicit.
```suggestion
  * When ChargeCategory is "Usage" or "Purchase", and when CommitmentDiscountId is not null, CommitmentDiscountQuantity adheres to the following additional requirements:
```

Reviewer: We agreed on this.  Folded this edit into the other suggestions; resolving.

    * CommitmentDiscountQuantity MUST NOT be null when [ChargeClass](#chargeclass) is not "Correction".
    * CommitmentDiscountQuantity MAY be null when ChargeClass is "Correction".
  * CommitmentDiscountQuantity MUST be null in all other cases.
* When CommitmentDiscountQuantity is not null, CommitmentDiscountQuantity adheres to the following additional requirements:
  * CommitmentDiscountQuantity MUST be a valid decimal value.
  * When ChargeCategory is "Purchase", CommitmentDiscountQuantity adheres to the following additional requirements:

>     * CommitmentDiscountQuantity MUST be the quantity of CommitmentDiscountUnits, paid fully or partially upfront, that is eligible for consumption over the *commitment discount's* *term* when [ChargeFrequency](#chargefrequency) is "One-Time".

Reviewer: There are multiple of these to apply in this section of requirements.

```suggestion
    * CommitmentDiscountQuantity MUST be the quantity of CommitmentDiscountUnit, paid fully or partially upfront, that is eligible for consumption over the *commitment discount's* *term* when [ChargeFrequency](#chargefrequency) is "One-Time".
```

    * CommitmentDiscountQuantity MUST be the quantity of CommitmentDiscountUnits that is eligible for consumption for each *charge period* that corresponds with the purchase when ChargeFrequency is "Recurring".
  * When ChargeCategory is "Usage", CommitmentDiscountQuantity adheres to the following additional requirements:
    * CommitmentDiscountQuantity MUST be the metered quantity of CommitmentDiscountUnits that is consumed over the *row's* *charge period* when [CommitmentDiscountStatus](#commitmentdiscountstatus) is "Used".

>     * CommitmentDiscountQuantity MUST be the remaining, unused quantity of CommitmentDiscountUnits for the *row's* *charge period* when CommitmentDiscountStatus is "Unused".

Reviewer: It appears that you dropped all references to the term `positive` from the 1.1 definition.  Why is this?

Author: As we established while discussing Correction handling, both positive and negative values are allowed (especially when coming from providers who do not restate the data but provide it in ledger style).

Reviewer: Discussed and agreed this is a bugfix.  This is an example of a change that would be helpful to see in a list of changes that are not strictly normative revisions.  @ijurica has such a list and shall add it to the PR / changelog.

Reviewer: Add to mergeable PR as a comment by tomorrow (Feb 25).

Author: added as a comment in PR #798

* When ChargeCategory is not "Usage" or "Purchase", EffectiveCost adheres to the following additional requirements:
  * EffectiveCost of a charge calculated based on other charges (e.g., when the ChargeCategory is "Tax") MUST be calculated based on the EffectiveCost of those related charges.
  * EffectiveCost of a charge unrelated to other charges (e.g., when the ChargeCategory is "Credit") MUST match the [BilledCost](#billedcost).
* When CommitmentDiscountStatus is "Unused", the EffectiveCost MUST be the total committed cost consumed for the given charge period minus related usage charges.

### **Effective Cost v.1.2 (Technical Refinement)**

The EffectiveCost column adheres to the following requirements:

* EffectiveCost MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* EffectiveCost MUST be of type Decimal.
* EffectiveCost MUST conform to [Numeric Format](#numericformat) requirements.
* EffectiveCost MUST NOT be null.
* EffectiveCost MUST be a valid decimal value.
* EffectiveCost MUST be 0 if [ChargeCategory](#chargecategory) is "Purchase" and the purchase is intended to cover future eligible charges.
* EffectiveCost MUST be denominated in the BillingCurrency.

> * The sum of EffectiveCost for *rows* in a given *billing period* may not match the sum of the invoices received for the same *billing period* for a [*billing account*](#glossary:billing-account).

Reviewer: Why did you change the phrasing from the 1.1 definition?  Is this part of a 1.2 Work Item?  Would like to hear the rationale.

Reviewer: Discussed and agreed this is capturing the intention of the previous definition.

> * The sum of EffectiveCost for *rows* in a given *billing period* may not match the sum of the invoices received for the same *billing period* for a [*billing account*](#glossary:billing-account).

Reviewer: Is this the right place in the order to put this?  I checked the cookbook, which doesn't match the definition proposed here, so I wasn't sure.  Could you lay out in a comment the flow for each section here, so I understand how it aligns with the guidelines?

Reviewer: Discussed and agreed this does follow the order of the cookbook.  Resolving.

> * The sum of EffectiveCost for *rows* in a given *billing period* may not match the sum of the invoices received for the same *billing period* for a [*billing account*](#glossary:billing-account).

Reviewer: Shall we strike the term `row` or shall we continue to use or shall we use a different term like `charge` or `line item`?

```suggestion
* The sum of EffectiveCost in a given *billing period* may not match the sum of the invoices received for the same *billing period* for a [*billing account*](#glossary:billing-account).
```

Author: I plan to apply the pattern 'The sum of ColumnId in a given billing period...' to the following requirements:

- The sum of the BilledCost for rows in a given billing period MUST match the sum of the invoices received for that billing period for a billing account.
- The sum of the BilledCost for rows in a given billing period MUST match the sum of the invoices received for that billing period for a billing account.
- The sum of the BilledCost for rows in a given billing period MUST match the sum of the invoices received for that billing period for a billing account.
- The sum of EffectiveCost for rows in a given billing period may not match the sum of the invoices received for the same billing period for a billing account.

Author: We could apply a similar change here as well:

Current version:

> - CommitmentDiscountQuantity MUST be the metered quantity of CommitmentDiscountUnit that is consumed over the row's charge period when CommitmentDiscountStatus is "Used".
> - CommitmentDiscountQuantity MUST be the remaining, unused quantity of CommitmentDiscountUnit for the row's charge period when CommitmentDiscountStatus is "Unused".

Proposed change:

> - CommitmentDiscountQuantity MUST represent the quantity of CommitmentDiscountUnit consumed in a given charge period when CommitmentDiscountStatus is 'Used'.
> - CommitmentDiscountQuantity MUST represent the remaining quantity of CommitmentDiscountUnit in a given charge period when CommitmentDiscountStatus is 'Unused'.

Or is it too much?

Reviewer: Discussed.  Will sustain the change eliminating `for rows`.  This needs to be tracked in a list of global changes alongside all others that can be explained to the working group.

We will not remove "metered".  We will not remove "unused".  We will not change "be" to "represent".

Author: agreed changes applied.

In cases where the [ChargeCategory](#chargecategory) is not "Usage" or "Purchase", the following applies:

* The EffectiveCost MUST be calculated based on the EffectiveCost of the related charges if the charge is calculated based on other charges (e.g. [ChargeCategory](#chargecategory) is "Tax").
* The EffectiveCost MUST match the [BilledCost](#billedcost) if the charge is unrelated to other charges (e.g. [ChargeCategory](#chargecategory) is "Credit").
* When CommitmentDiscountStatus is "Unused", the EffectiveCost MUST be the total committed cost consumed for the given charge period minus related usage charges.

## Column: Invoice Issuer

## Column: List Cost

### **List Cost v.1.2 (Simplified Refinement)**

The ListCost column adheres to the following requirements:

* ListCost MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ListCost MUST be of type Decimal.
* ListCost MUST conform to [Numeric Format](#numericformat) requirements.
* ListCost MUST NOT be null.
* ListCost MUST be a valid decimal value.
* ListCost MUST be denominated in the BillingCurrency.

> * When [ListUnitPrice](#listunitprice) is null, ListCost adheres to the following additional requirements:
>   * ListCost of a charge calculated based on other charges (e.g., when the [ChargeCategory](#chargecategory) is "Tax") MUST be calculated based on the ListCost of those related charges.
>   * ListCost of a charge unrelated to other charges (e.g., when the ChargeCategory is "Credit") MUST match the [BilledCost](#billedcost).

Reviewer: The clause `When ListUnitPrice is null` should be `not null` instead -- but actually, because the column MUST NOT be null as stated earlier, we can skip this conditional completely, as far as I can tell.
```suggestion
* ListCost of a charge calculated based on other charges (e.g., when the [ChargeCategory](#chargecategory) is "Tax") MUST be calculated based on the ListCost of those related charges.
* ListCost of a charge unrelated to other charges (e.g., when the ChargeCategory is "Credit") MUST match the [BilledCost](#billedcost).
```

* The product of ListUnitPrice and PricingQuantity MUST match the ListCost when ListUnitPrice is not null, PricingQuantity is not null, and [ChargeClass](#chargeclass) is not "Correction".
* Discrepancies in ListCost, ListUnitPrice, or PricingQuantity MAY be addressed independently when ChargeClass is "Correction".

### **List Cost v.1.2 (Technical Refinement)**

The ListCost column adheres to the following requirements:

* ListCost MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ListCost MUST be of type Decimal.
* ListCost MUST conform to [Numeric Format](#numericformat) requirements.
* ListCost MUST NOT be null.
* ListCost MUST be a valid decimal value.
* ListCost MUST be denominated in the BillingCurrency.

> * If [ListUnitPrice](#listunitprice) is present and null, ListCost adheres to the following additional requirements:

Reviewer: Simple error, simple fix.

```suggestion
* If [ListUnitPrice](#listunitprice) is present and not null, ListCost adheres to the following additional requirements:
```

* This column MUST be a Decimal within the range of non-negative decimal values, MUST conform to [Numeric Format](#numericformat) requirements, and be denominated in the BillingCurrency.
* It MUST NOT be null when [ChargeClass](#chargeclass) is not "Correction" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.
* When ListUnitPrice is present and is not null, multiplying ListUnitPrice by [PricingQuantity](#pricingquantity) MUST equal [ListCost](#listcost), except in cases of ChargeClass "Correction", which may address PricingQuantity or any cost discrepancies independently.

## Column: Pricing Category
## Column: Pricing Quantity
## Column: Pricing Unit
## Column: Provider
## Column: Publisher
## Column: Region ID
## Column: Region Name
## Column: Resource ID
## Column: Resource Name
## Column: Resource Type
## Column: Service Category
## Column: Service Name
## Column: Service Subcategory
## Column: SKU ID
## Column: SKU Meter

### **SKU Meter v.1.2 (Simplified Refinement)**

The SkuMeter column adheres to the following requirements:

> * SkuMeter MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider publishes a SKU list.

Reviewer: Why did we make this change from `SkuId` to `SKU list`?  And if we're going to stick with this change: do we need an entry in the glossary?

Author: The guideline to use standardized phrasing and terminology in the context of column presence conditions could (or should) be interpreted as using the same wording when the same condition applies. In line with this, I suggest we use the same wording for:

- SkuId and SkuMeter column presence conditions, i.e., 'when the provider publishes a SKU list'.
- SkuPriceId and SkuPriceDetails, i.e., 'when the provider publishes a SKU price list'.

Additionally, I believe that we should use the same presence condition for SkuId, SkuPriceId, and the columns related to them. However, this is a discussion topic for WorkItem #314

Reviewer: Discussed.  Agreed that "publishes a SKU list" is the same requirement as when `SkuId` is present, so it's the same requirement that can be repeated across multiple columns when analyzing normative requirements.  To the second part of the question, what is a SKU list: this is something worth discussing but beyond the scope of this exercise.

* SkuMeter MUST be of type String.
* SkuMeter MUST conform to [String Handling](#stringhandling) requirements.
* SkuMeter nullability is defined as follows:

>   * SkuMeter MUST be null if SkuId is null.

Reviewer: This is the first reference to SkuId, right?  If so:
```suggestion
  * SkuMeter MUST be null if [SkuId](#skuid) is null.
```

Reviewer: Irena to make change in mergeable PR by tomorrow (Feb 25).

  * SkuMeter SHOULD NOT be null if SkuId is not null.

> * SkuMeter SHOULD remain consistent over time for a given SkuId.

Author: Skipped 'when SkuMeter is not null' as it sounds awkward and felt unnecessary

Reviewer: Agreed!

The SkuMeter column adheres to the following requirements:

* SkuMeter MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider publishes a SKU list.
* If present, SkuMeter adheres to the following additional requirements:
  * SkuMeter MUST be of type String.
  * SkuMeter MUST conform to [String Handling](#stringhandling) requirements.
  * SkuMeter MUST be null if SkuId is null.
  * SkuMeter SHOULD NOT be null if SkuId is not null.
  * If SkuMeter is not null, SkuMeter adheres to the following additional requirement:
    * SkuMeter SHOULD remain consistent over time for a given SkuId.

### **SKU Meter v.1.1 (Original)**

The SkuMeter column adheres to the following requirements:

* SkuMeter MUST be present in a *FOCUS dataset* when when the provider includes a [SkuId](#skuid).
* SkuMeter MUST be of type String.
* SkuMeter MUST be null when SkuId is null.
* SkuMeter SHOULD NOT be null when SkuId is not null.
* SkuMeter SHOULD remain consistent over time for a given SkuId.

## Column: SKU Price Details

### **SKU Price Details v.1.2 (Simplified Refinement)**

The SkuPriceDetails column adheres to the following requirements:

> * SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider publishes a SKU price list.

Reviewer: Same note as SkuMeter.  Why have we switched from a reference to `SkuPriceId` to `SKU price list`?

If this is the way forward, then the first reference to `SkuPriceId` later needs a link.

Author: see comment https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/664#discussion_r1959761160

Reviewer: Discussed.  Same outcome as for SkuMeter.  Resolving.

* SkuPriceDetails MUST conform to [KeyValueFormat](#key-valueformat) requirements.

> * SkuPriceDetails property key SHOULD conform to [PascalCase](#glossary:pascalcase) format.

Reviewer: Maybe we could skip the word `property`?  Or is that critical?

```suggestion
* SkuPriceDetails keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
```

Reviewer: If that's not feasible, then I think it reads better if it's plural...?

```suggestion
  * SkuPriceDetails property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
```

Author: My initial reasoning was that SkuPriceDetails contains a list of properties (property keys and values), while the Tags column contains a list of tags (tag keys and values), which is why I used the terms 'property key' and 'tag key'. 

I do like the suggestion to simplify and just use 'key' and 'value' for both SkuPriceDetails and Tags.

Reviewer: Discussed; aligned on first option.

Author: Alternative version provided in PR #798. Reasoning is outlined in the guidelines (see PR #787)

* SkuPriceDetails nullability is defined as follows:
  * SkuPriceDetails MUST be null when SkuPriceId is null.
  * SkuPriceDetails MAY be null when SkuPriceId is not null.
* When SkuPriceDetails is not null, SkuPriceDetails adheres to the following additional requirements:
  * SkuPriceDetails MUST NOT contain properties which are not applicable to the corresponding SkuPriceId.
  * SkuPriceDetails MAY contain properties which are already captured in other dedicated columns.
  * SkuPriceDetails SHOULD remain consistent over time for a given SkuPriceId.
  * SkuPriceDetails properties for a given SkuPriceId adhere to the following additional requirements:

>     * Existing properties SHOULD remain consistent over time.
>     * Existing properties SHOULD NOT be removed.
>     * Additional properties MAY be added over time.

Reviewer: One of our principles is to stick to starting with the column name, and I don't see a good enough reason to avoid that here.  So, how about:

```suggestion
      * SkuPriceDetails existing properties SHOULD remain consistent over time.
      * SkuPriceDetails existing properties SHOULD NOT be removed.
      * SkuPriceDetails additional properties MAY be added over time.
```

Reviewer: If that's too awkward, how about:

```suggestion
      * Existing SkuPriceDetails properties SHOULD remain consistent over time.
      * Existing SkuPriceDetails properties SHOULD NOT be removed.
      * Additional SkuPriceDetails properties MAY be added over time.
```

Reviewer: Discussed and agreed to use the second option.

  * SkuPriceDetails property key SHOULD remain consistent across comparable SKUs having that property, and the values for this key SHOULD remain in a consistent format.
  * SkuPriceDetails property value MUST represent the value for a single [PricingUnit](#pricingunit) when the property holds a numeric value.

### **SKU Price Details v.1.2 (Technical Refinement)**

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider publishes a SKU price list.
* If present, SkuPriceDetails adheres to the following additional requirements:
  * SkuPriceDetails MUST conform to [KeyValueFormat](#key-valueformat) requirements.

>   * SkuPriceDetails property key SHOULD conform to [PascalCase](#glossary:pascalcase) format.

Reviewer: I understand that this follows the pattern we've established... but it reads awkwardly.  How can we finesse it and still stick to the principle of starting with the word `SkuPriceDetails`?

Reviewer: Maybe we could skip the word `property`?  Or is that critical?

```suggestion
  * SkuPriceDetails keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
```

Reviewer: If that's not feasible, then I think it reads better if it's plural...?

```suggestion
  * SkuPriceDetails property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
```

  * SkuPriceDetails MUST be null when SkuPriceId is null.
  * SkuPriceDetails MAY be null when SkuPriceId is not null.
  * If SkuPriceDetails is not null, SkuPriceDetails adheres to the following additional requirements:
    * SkuPriceDetails MUST NOT contain properties which are not applicable to the corresponding SkuPriceId.
    * SkuPriceDetails MAY contain properties which are already captured in other dedicated columns.
    * SkuPriceDetails SHOULD remain consistent over time for a given SkuPriceId.
    * SkuPriceDetails properties for a given SkuPriceId adhere to the following additional requirements:

>       * Existing properties SHOULD remain consistent over time.
>       * Existing properties SHOULD NOT be removed.
>       * Additional properties MAY be added over time.

Reviewer: One of our principles is to stick to starting with the column name, and I don't see a good enough reason to avoid that here.  So, how about:

```suggestion
      * SkuPriceDetails existing properties SHOULD remain consistent over time.
      * SkuPriceDetails existing properties SHOULD NOT be removed.
      * SkuPriceDetails additional properties MAY be added over time.
```

Reviewer: If that's too awkward, how about:

```suggestion
      * Existing SkuPriceDetails properties SHOULD remain consistent over time.
      * Existing SkuPriceDetails properties SHOULD NOT be removed.
      * Additional SkuPriceDetails properties MAY be added over time.
```

* SkuPriceId MUST NOT be null when [ChargeClass](#chargeclass) is not "Correction" and [ChargeCategory](#chargecategory) is "Usage" or "Purchase", MUST be null when ChargeCategory is "Tax", and MAY be null for all other combinations of ChargeClass and ChargeCategory.
* A given value of SkuPriceId MUST be associated with one and only one [SkuId](#skuid), except in cases of [commitment discount flexibility](#glossary:commitment-discount-flexibility).
* If a provider does not have a SkuPriceId and wants to include information in columns linked to SkuPriceId such as ListUnitPrice or [SkuPriceDetails](#skupricedetails), the SkuId MAY be used in the SkuPriceId column as long as it adheres to the above conditions.

## Column: Sub Account ID

### **Sub Account ID v.1.2 (Simplified Refinement)**

The SubAccountId column adheres to the following requirements:

* SubAccountId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports a *sub account* construct.
* SubAccountId MUST be of type String.
* SubAccountId MUST conform to [String Handling](#stringhandling) requirements.
* SubAccountId nullability is defined as follows:

>   * SubAccountId MUST be null when a charge is not related to a *sub account*.
>   * SubAccountId MUST NOT be null when a charge is related to a *sub account*.

Reviewer: Not sure how we add it to the guidelines, but it feels like the importance of these is swapped.  Maybe we add a guideline that says, "Please order groups of requirements by perceived importance"?

```suggestion
  * SubAccountId MUST NOT be null when a charge is related to a *sub account*.
  * SubAccountId MUST be null when a charge is not related to a *sub account*.  
```

Reviewer: The way it is _does_ follow the guidelines, starting with `MUST` and following with `MUST NOT`.  However, maybe it could make more sense to start with `MUST NOT` and then follow with `MUST`.  @ijurica to analyze.

Author: @shawnalpay 
In the case of CommitmentDiscountStatus (see example), the MUST condition defines the primary rule, while MUST NOT addresses the exception. Given this, I prefer the existing order of conditions.

> * CommitmentDiscountStatus nullability is defined as follows:
>   * CommitmentDiscountStatus MUST be null when [CommitmentDiscountId](#commitmentdiscountid) is null.
>   * CommitmentDiscountStatus MUST NOT be null when CommitmentDiscountId is not null and [Charge Category](#chargecategory) is "Usage".

I reviewed all 50 columns and found a similar pattern in the following columns:
- CommitmentDiscountStatus, 
- ContractedUnitPrice, 
- ListUnitPrice, 
- PricingCategory, 
- PricingQuantity, 
- PricingUnit, 
- SkuId

Changing the prescribed condition order would turn these instances into exceptions to the guideline. Given the number of instances, **I propose keeping the order AS-IS.**

* SubAccountName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports a *sub account* construct.
* If present, SubAccountName adheres to the following additional requirements:
  * SubAccountName MUST be of type String.
  * SubAccountName MUST conform to [String Handling](#stringhandling) requirements.
  * SubAccountName MUST be null if [SubAccountId](#subaccountid) is null.
  * SubAccountName MUST NOT be null if SubAccountId is not null.

### **Sub Account Name v.1.1 (Original)**

The SubAccountName column adheres to the following requirements:

* The SubAccountName column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports a *sub account* construct.
* This column MUST be of type String.
* If a charge does not apply to a *sub account*, the SubAccountName column MUST be null.

## Column: Tags

### **Tags v.1.2 (Simplified Refinement)**

The Tags column adheres to the following requirements:

* Tags MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports setting user or provider-defined tags.
* Tags MUST conform to [KeyValueFormat](#key-valueformat) requirements.

> * Tags MAY be null.

Author: Decided to include the nullability-related requirement ('Tags MAY be null.') after all for consistency

Reviewer: Agreed.  This is the only column that didn't have nullability explicitly declared.

* When Tags is not null, Tags adheres to the following additional requirements:
  * Tags MUST contain user-defined and provider-defined tags.
  * Tags MUST only contain finalized tags.

>   * Tag key with a non-null value for a given resource SHOULD be included in the Tags column.

Reviewer: I feel like we should either keep starting with the column name, or keep it grammatically correct, the way we did in `SkuPriceDetails`.  Something like:

```suggestion
    * Tags SHOULD include a key with a non-null value for a given resource.
```

Reviewer: @cnharris10 Given your background on Tags, how can we restate this to potentially start with the word `Tags` to follow the guideline of always trying to start with the column name?

Reviewer: Irena to copy comment on mergeable PR by tomorrow (Feb 25).

Author: link to comment copied to the PR #798

>   * Tag key with a non-null value for a given resource SHOULD be included in the Tags column.

Reviewer: Should we make a glossary reference for `resource`?

Reviewer: Discussed; Irena to add.

Author: The column already contains a reference, so I've just added italics.

>   * Tag key with a null value for a given resource MAY be included in the Tags column depending on the provider's tag finalization process.

Reviewer: SkuPriceDetails made use of the term `property key`.  We should either use that term here, or simplify it there.  I think just using `key` is fine, personally.

Reviewer: Discussed.  Irena to revise.

Reviewer: Discussed.  Irena to include suggestion or propose alternative by tomorrow (Feb 25) in mergeable PR (part 2).

Author: Alternative version provided in PR #798. Reasoning is outlined in the guidelines (see PR #787).

  * Tag key that does *not* support a corresponding value, MUST have a corresponding true (boolean) value set.

>   * Providers MUST publish tag finalization methods and semantics within their respective documentation when tag finalization is supported.

Reviewer: Given the earlier requirement of `Tags MUST only contain finalized tags`, why do we have this `when tag finalization is supported` here?  I know that's the way it's written in 1.1, but that strikes me as strange.  Shall we consider:

```suggestion
  * Providers MUST publish tag finalization methods and semantics within their documentation.
```

Reviewer: @cnharris10 Do you agree with my point above, or is this too much changing the definition?

Reviewer: Irena will copy this comment into the mergeable PR by tomorrow (Feb 25).

  * Providers MUST NOT alter user-defined tag keys or values.
  * Provider-defined tags adhere to the following additional requirements:
    * Provider-defined tags MUST be prefixed with a provider-specified tag key prefix.
    * Providers SHOULD publish all provider-specified tag key prefixes within their respective documentation.

### **Tags v.1.2 (Technical Refinement)**

The Tags column adheres to the following requirements:

* Tags MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports setting user or provider-defined tags.
* If present, Tags adheres to the following additional requirements:
  * Tags MUST conform to [KeyValueFormat](#key-valueformat) requirements.
  * If Tags is not null, Tags adheres to the following additional requirements:
    * Tags MUST contain user-defined and provider-defined tags.
    * Tags MUST only contain finalized tags.

>     * Tag key with a non-null value for a given resource SHOULD be included in the Tags column.

Reviewer: I feel like we should either keep starting with the column name, or keep it grammatically correct, the way we did in `SkuPriceDetails`.  Something like:

```suggestion
    * Tags SHOULD include a key with a non-null values for a given resource.
```

### **Billed Cost v.1.1 (Original)**

The BilledCost column adheres to the following requirements:

* The BilledCost column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) and MUST NOT be null.
* This column MUST be of type Decimal, MUST conform to [Numeric Format](#numericformat), and be denominated in the BillingCurrency.
* The sum of the BilledCost for [*rows*](#glossary:row) in a given [*billing period*](#glossary:billing-period) MUST match the sum of the invoices received for that *billing period* for a [*billing account*](#glossary:billing-account).

## Column: Billing Account ID

### **Billing Account ID v.1.2 (Simplified Refinement)**

The BillingAccountId column adheres to the following requirements:

* BillingAccountId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* BillingAccountId MUST be of type String.
* BillingAccountId MUST conform to [String Handling](#stringhandling) requirements.
* BillingAccountId nullability is defined as follows:
  * BillingAccountId MUST NOT be null.

> * BillingAccountId MUST ensure global uniqueness within a provider.

Reviewer: ```suggestion
* BillingAccountId MUST be a unique identifier within a provider.
```

Author: addressed

* BillingAccountId MUST NOT be null.
* BillingAccountId MUST be a globally unique identifier within a provider.

### **Billing Account ID v.1.1 (Original)**

The BillingAccountId column adheres to the following requirements:

* The BillingAccountId column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* This column MUST be of type String and MUST NOT contain null values.
* BillingAccountId MUST be a globally unique identifier within a provider.

## Column: Billing Account Name

### **Billing Account Name v.1.2 (Simplified Refinement)**

The BillingAccountName column adheres to the following requirements:

* BillingAccountName MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* BillingAccountName MUST be of type String.
* BillingAccountName MUST conform to [String Handling](#stringhandling) requirements.
* BillingAccountName nullability is defined as follows:
  * BillingAccountName MUST NOT be null when the provider supports assigning a display name for the *billing account*.
* When BillingAccountName is not null, BillingAccountName adheres to the following additional requirement:

>   * BillingAccountName MUST ensure uniqueness within a customer.

Reviewer: Global application of "be unique" to replace "ensure uniqueness".

```suggestion
  * BillingAccountName MUST be unique within a customer.
```

Author: addressed

* If a property has a numeric value, it MUST represent the value for a single [PricingUnit](#pricingunit).
* The SkuPriceDetails column MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider includes a SkuPriceId.
  * The SkuPriceDetails column MAY be null when SkuPriceId is not null.
  * The SkuPriceDetails column MUST be null when SkuPriceId is null.

## Column: SKU Price ID

### **SKU Price ID v.1.2 (Simplified Refinement)**

The SkuPriceId column adheres to the following requirements:

* SkuPriceId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider publishes a SKU price list.
* SkuPriceId MUST be of type String.
* SkuPriceId MUST conform to [String Handling](#stringhandling) requirements.
* SkuPriceId nullability is defined as follows:
  * SkuPriceId MUST be null when [ChargeCategory](#chargecategory) is "Tax".
  * SkuPriceId MUST NOT be null when ChargeCategory is "Usage" or "Purchase" and [ChargeClass](#chargeclass) is not "Correction".
  * SkuPriceId MAY be null in all other cases.
* When SkuPriceId is not null, SkuPriceId adheres to the following additional requirements:

>   * SkuPriceId MUST be associated with one and only one [SkuId](#skuid), except in cases of [commitment discount flexibility](#glossary:commitment-discount-flexibility).
>   * SkuPriceId MUST define a single unit price used for calculating the charge.
>   * When a provider does not have a SkuPriceId and wants to include information in columns linked to SkuPriceId such as ListUnitPrice or [SkuPriceDetails](#skupricedetails), the SkuId MAY be used in the SkuPriceId column as long as it adheres to the above conditions.
>   * [ListUnitPrice](#listunitprice) MUST be associated with the SkuPriceId in the provider published price list.

Author: To be revisited as part of PR #676 

```suggestion
  * SkuPriceId MUST be associated with one and only one [SkuId](#skuid), except in cases of [commitment discount flexibility](#glossary:commitment-discount-flexibility).
  * SkuPriceId MUST define a single unit price used for calculating the charge.
  * When a provider does not have a SkuPriceId and wants to include information in columns linked to SkuPriceId such as ListUnitPrice or [SkuPriceDetails](#skupricedetails), the SkuId MAY be used in the SkuPriceId column as long as it adheres to the above conditions.
  * [ListUnitPrice](#listunitprice) MUST be associated with the SkuPriceId in the provider published price list.
```

Reviewer: Irena to place in list of discussion topics per AI https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/719.
